crates/solver/src/s3_instance_upload_arguments.rs

```diff
@@ -45,6 +31,53 @@ impl std::fmt::Display for S3UploadArguments {
             "s3_instance_upload_filename_prefix",
             &self.s3_instance_upload_filename_prefix,
         )?;
+        Ok(())
+    }
+}
+
+impl S3UploadArguments {
+    pub fn into_config(self) -> anyhow::Result<Option<Config>> {
+        let s3_args = &[
+            &self.s3_instance_upload_bucket,
+            &self.s3_instance_upload_filename_prefix,
+        ];
+        let all_some = s3_args.iter().all(|arg| arg.is_some());
+        let all_none = s3_args.iter().all(|arg| arg.is_none());
+        anyhow::ensure!(
+            all_some || all_none,
+            "either set all s3_instance_upload bucket arguments or none"
+        );
+        Ok(if all_some {
+            Some(Config {
+                bucket: self.s3_instance_upload_bucket.unwrap(),
+                filename_prefix: self.s3_instance_upload_filename_prefix.unwrap(),
+            })
+        } else {
+            None
+        })
+    }
+}
+
+#[derive(clap::Parser)]
+pub struct S3CredentialArguments {
```

Comment:
nit: Are these credentials really local to S3 uploads or more generally the AWS credentials for this pod. Like if we had another service that would require auth would it even be possible to use different credentials (e.g. in the IRSA case)?

Reply:
The CLI arguments are specific to S3. So theoretically if we had another use case for AWS credentials we could introduce another set of CLI arguments for that specifically so they are completely independent from each other.
AFAIK this is not the case for IRSA. The deal with that seems to be that pod is like "Hey AWS I'm pod X, please make sure I'm only allowed to do stuff devops permitted me to do".
For that mode of operations it doesn't really make sense to allow multiple different IRSA credentials (just set up one and make sure it has all the permissions it needs).

Overall I would be in favor of ditching the CLI arguments ASAP. If we are able to set up all the IRSA config stuff before merging this PR we can already get rid of that in here.

Reply:
@mateipopa it would be possible to make our infra completely IRSA ready before merging this PR so we can remove the CLI argument already, right?

Reply:
Yes, that's the plan. I'll ping as soon as I'm ready with the infra